File: src/main/java/net/treset/minecraftlauncher/creation/JavaComponentCreator.java
package net.treset.minecraftlauncher.creation;

import net.treset.mc_version_loader.exception.FileDownloadException;
import net.treset.mc_version_loader.java.*;
import net.treset.mc_version_loader.launcher.LauncherManifest;
import net.treset.mc_version_loader.launcher.LauncherManifestType;
import net.treset.mc_version_loader.util.OsUtil;
import net.treset.minecraftlauncher.util.CreationStatus;
import net.treset.minecraftlauncher.util.exception.ComponentCreationException;

import java.io.File;
import java.util.List;
import java.util.Map;

public class JavaComponentCreator extends GenericComponentCreator {
    public JavaComponentCreator(String name, Map<String, LauncherManifestType> typeConversion, LauncherManifest componentsManifest) {
        super(LauncherManifestType.JAVA_COMPONENT, null, null, name, typeConversion, null, null, componentsManifest);
        setDefaultStatus(new CreationStatus(CreationStatus.DownloadStep.JAVA, null));
    }

    @Override
    public String createComponent() throws ComponentCreationException {
        String result = super.createComponent();

        if(result == null || getNewManifest() == null) {
            attemptCleanup();
            throw new ComponentCreationException("Failed to create java component: invalid data");
        }

        JavaRuntimes java;
        try {
            java = JavaUtil.getJavaRuntimes();
        } catch (FileDownloadException e) {
            attemptCleanup();
            throw new ComponentCreationException("Failed to create java component: failed to download java runtime json", e);
        }

        String osIdentifier;
        try {
            osIdentifier = OsUtil.getJavaIdentifier();
        } catch (IllegalArgumentException e) {
            attemptCleanup();
            throw new ComponentCreationException("Failed to create java component: failed to get os identifier", e);
        }

        JavaRuntimeOs os = null;
        for(JavaRuntimeOs o : java.getRuntimes()) {
            if(o.getId().equals(osIdentifier)) {
                os = o;
                break;
            }
        }
        if(os == null || os.getReleases() == null) {
            attemptCleanup();
            throw new ComponentCreationException("Failed to create java component: failed to get os runtime");
        }

        JavaRuntimeRelease release = null;
        for(JavaRuntimeRelease r : os.getReleases()) {
            if(r != null && getName().equals(r.getId())) {
                release = r;
                break;
            }
        }

        if(release == null || release.getManifest() == null || release.getManifest().getUrl() == null) {
            attemptCleanup();
            throw new ComponentCreationException("Failed to create java component: failed to get release");
        }

        List<JavaFile> files;
        try {
            files = JavaUtil.getJavaFile(release.getManifest().getUrl());
        } catch (FileDownloadException e) {
            throw new ComponentCreationException("Failed to create java component: failed to download java file manifest", e);
        }

        File baseDir = new File(getNewManifest().getDirectory());

        try {
            JavaUtil.downloadJavaFiles(baseDir, files, status -> setStatus(new CreationStatus(CreationStatus.DownloadStep.JAVA, status)));
        } catch (FileDownloadException e) {
            throw new ComponentCreationException("Failed to create java component: failed to download java files", e);
        }

        return result;
    }

    @Override
    public String useComponent() throws ComponentCreationException {
        attemptCleanup();
        throw new ComponentCreationException("Unable to use java component: unsupported");
    }

    @Override
    public String inheritComponent() throws ComponentCreationException {
        attemptCleanup();
        throw new ComponentCreationException("Unable to inherit java component: unsupported");
    }
}

File: client/composeApp/src/commonMain/kotlin/net/treset/treelauncher/backend/creation/JavaComponentCreator.kt
package net.treset.treelauncher.backend.creation

import net.treset.mc_version_loader.exception.FileDownloadException
import net.treset.mc_version_loader.java.*
import net.treset.mc_version_loader.util.DownloadStatus
import net.treset.mc_version_loader.util.OsUtil
import net.treset.treelauncher.backend.data.manifest.LauncherManifestType
import net.treset.treelauncher.backend.data.manifest.ParentManifest
import net.treset.treelauncher.backend.util.CreationStatus
import net.treset.treelauncher.backend.util.exception.ComponentCreationException
import net.treset.treelauncher.backend.util.file.LauncherFile

class JavaComponentCreator(
    name: String,
    typeConversion: Map<String, LauncherManifestType>,
    componentsManifest: ParentManifest
) : GenericComponentCreator(
    LauncherManifestType.JAVA_COMPONENT,
    null,
    null,
    name,
    typeConversion,
    null,
    null,
    componentsManifest
) {
    init {
        defaultStatus = CreationStatus(CreationStatus.DownloadStep.JAVA, null)
    }

    @Throws(ComponentCreationException::class)
    override fun createComponent(): String {
        val result = super.createComponent()
        if (newManifest == null) {
            attemptCleanup()
            throw ComponentCreationException("Failed to create java component: invalid data")
        }
        val java: JavaRuntimes = try {
            MinecraftJava.getJavaRuntimes()
        } catch (e: FileDownloadException) {
            attemptCleanup()
            throw ComponentCreationException("Failed to create java component: failed to download java runtime json", e)
        }
        val osIdentifier: String = try {
            OsUtil.getJavaIdentifier()
        } catch (e: IllegalArgumentException) {
            attemptCleanup()
            throw ComponentCreationException("Failed to create java component: failed to get os identifier", e)
        }
        val os = java[osIdentifier]
        if (os?.entries == null || os.entries.isEmpty()) {
            attemptCleanup()
            throw ComponentCreationException("Failed to create java component: failed to get os runtime")
        }

        val release = os[name]?.get(0)
        if (release?.manifest == null || release.manifest.url == null) {
            attemptCleanup()
            throw ComponentCreationException("Failed to create java component: failed to get release")
        }
        val files: List<JavaFile> = try {
            MinecraftJava.getJavaFiles(release.manifest.url)
        } catch (e: FileDownloadException) {
            throw ComponentCreationException("Failed to create java component: failed to download java file manifest", e)
        }
        val baseDir = LauncherFile(newManifest!!.directory)
        try {
            MinecraftJava.downloadJavaFiles(
                baseDir,
                files
            ) { status: DownloadStatus? ->
                setStatus(
                    CreationStatus(
                        CreationStatus.DownloadStep.JAVA,
                        status
                    )
                )
            }
        } catch (e: FileDownloadException) {
            throw ComponentCreationException("Failed to create java component: failed to download java files", e)
        }
        return result
    }

    @Throws(ComponentCreationException::class)
    override fun useComponent(): String {
        attemptCleanup()
        throw ComponentCreationException("Unable to use java component: unsupported")
    }

    @Throws(ComponentCreationException::class)
    override fun inheritComponent(): String {
        attemptCleanup()
        throw ComponentCreationException("Unable to inherit java component: unsupported")
    }
}
